Improve checking dummy kernel module

StefanScherer opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**

Improve the commit https://github.com/docker/docker/commit/f3b2233d126aec8ab15fa589f5a9360eec280f01 to work without root user on an ARM machine. Just check if the kernel module "dummy" is already loaded. Only if not show a warning and try to run the commands to load it (but this requires root access). Added some notes in the ARM.md as well.

**\- How I did it**

**\- How to verify it**

Create a Scaleway C1 machine with Docker image.

Run `make test-integration-cli` on a Scaleway machine as non-root user. It aborts as the  `dummy` kernel module is not loaded.

Run these commands as root user:

```
modprobe dummy
echo "dummy" >>/etc/modules
```

You now may reboot to test automatically loading the kernel image.

Run `make test-integration-cli` again as non-root user and see that it now begins to build.

closes #20793

StefanScherer committed:
Improve checking dummy kernel module

Signed-off-by: Stefan Scherer <scherer_stefan@icloud.com>

calavera commented:
cc @tianon, @jfrazelle

jfrazelle commented:
LGTM

On Friday, March 11, 2016, David Calavera <notifications@github.com> wrote:

> cc @tianon <https://github.com/tianon>, @jfrazelle
> <https://github.com/jfrazelle>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/21110#issuecomment-195486654>.
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

tianon commented on Makefile:
```diff
@@ -85,14 +85,17 @@ binary: build
+	  echo "loaded. However, the dummy module is not loaded."; \
+	  echo "Please load the kernel module permanently."; \
+	  echo "Syncing kernel modules"; \
+	  oc-sync-kernel-modules; \
```
Does this tool exist in general ARM boxes? I'm not familiar with it. :confused:

StefanScherer replied:
I think this is a Scaleway only tool. I left that inside as it seems that the ARM builder needed that. But this can also be done while provisioning new machines in Scaleway.

tianon commented on Makefile:
```diff
@@ -85,14 +85,17 @@ binary: build
+	  oc-sync-kernel-modules; \
+	  depmod; \
+	  modprobe dummy; \
+	  echo "dummy" >> /etc/modules; \
```
Modifying /etc/modules on the host won't work without sudo, right? Also seems kind of dangerous to just append to this file unilaterally.

StefanScherer replied:
This also would be better while provisioning a Jenkins build slave.

StefanScherer commented:
I just have seen that the Jenkins configuration has all these steps for the Scaleway machines. So I suggest to remove the entire if block from the Makefile. WDYT?

tianon commented:
I'm +1!  Are they also documented somewhere so folks testing without our CI
can find them?

StefanScherer commented:
Haven't used this tool, but it is explained here to update kernel modules in Scaleway machines: https://www.scaleway.com/docs/how-your-kernel-works/

So this is specific to Scaleway and the use of should be removed from the general Makefile.

The "missing" dummy kernel module also not ARM specific. Just spun up a Vagrant box with boxcutter/ubuntu1404 and this kernel module also has to be loaded with modprobe.

I will update the PR by removing the if block from the Makefile.

tianon commented:
Cool, sounds great!  Probably want to prefix it with a hyphen so that we
continue if it fails, right?

StefanScherer commented:
I have updated the PR and removed the whole if block with all the commands from the Makefile.

I did a test on a Scaleway machine with and without the dummy kernel module loaded. Both integration tests show up with the same result `OK: 1193 passed, 50 skipped`. Both tests are started as a user jenkins without root privileges.

jacksgt commented:
+1, just encountered this issue on an ARM machine, too!

Good to see someone has already figured it out :-)

tianon commented:
LGTM!

thaJeztah commented:
@jfrazelle ptal

thaJeztah commented:
(PR was updated since your LGTM)

calavera commented:
LGTM
